src/asn_index.c

```diff
@@ -0,0 +1,256 @@
+extern char * geoip_asn_v6_dat;
+extern int geoip_asn_v6_options;
+static hashfunc asn_hashfunc;
+static hashkeycmp asn_cmpfunc;
+
+#define MAX_ARRAY_SZ 65536
+static hashtbl *theHash = NULL;
+static int next_idx = 0;
+static GeoIP *geoip = NULL;
+static GeoIP *geoip6 = NULL;
+static char *ipstr;
+static char *nodb = "noGeoIPDB";
+static char *unknown = "unknown";
+static char *unknown_v4 = "unknown IPv4";
+static char *unknown_v6 = "unknown IPv6";
+
+typedef struct {
+    char *asn;
+    int index;
+}      asnobj;
+
+char *
+asn_get_from_message(dns_message * m)
+{
+    transport_message *tm;
+    char *asn;
+    char *truncate;
+
+    tm = m->tm;
+
+    if (!inXaddr_ntop(&tm->src_ip_addr, ipstr, 80)) {
+        dfprint(0, "asn_index: Error converting IP address");
+        return(strdup(unknown));
+    }
+    dfprintf(0, "asn_index: IP %s is IPv%d", ipstr, tm->ip_version);
+
+    switch(tm->ip_version) {
+    case 4:
+        if (geoip) {
+            asn = GeoIP_name_by_addr(geoip,ipstr);
+            if (asn == NULL) {
+                asn = strdup(unknown_v4); /* strdup uses malloc, just like libgeoip */
+            }
+        } else {
+	        asn = strdup(nodb);
+        }
+        break;
+    case 6:
+        if (geoip6) {
+            asn = GeoIP_name_by_addr_v6(geoip6,ipstr);
+            if (asn == NULL) {
+                asn = strdup(unknown_v6); /* strdup uses malloc, just like libgeoip */
+            }
+            break;
+        } else {
+	        asn = strdup(nodb);
+        }
+    default:
+        asn = strdup(unknown);
+    }
+    if (asn == NULL) {
+        dfprint(0, "asn_index: Error allocating memory");
+        return(NULL);
+    }
+
+    dfprintf(0, "asn_index: full network name: %s", asn);
+
+    /* libgeoip reports for networks with the same ASN different network names.
+       Probably it uses the network description, not the AS description. Therefore,
+       we truncate after the first space and only use the AS number. Mappings
+       to AS names must be done in the presenter.
+    */
+    truncate = strchr(asn,' ');
+    if (truncate) {
+        *truncate='\0';
+    }
+
+    dfprintf(0, "asn_index: truncated network name: %s", asn);
+    return (asn);
+}
+
+int
+asn_indexer(const void *vp)
+{
+    const dns_message *m = vp;
+    char *asn;
+    asnobj *obj;
+    if (m->malformed)
+        return -1;
+    asn = asn_get_from_message((dns_message *) m);
+    if (asn == NULL)
+        return -1;
+    if (NULL == theHash) {
+        theHash = hash_create(MAX_ARRAY_SZ, asn_hashfunc, asn_cmpfunc,
+            1, afree, afree);
+        if (NULL == theHash) {
+            free(asn);
+            return -1;
+        }
+    }
+    if ((obj = hash_find(asn, theHash))) {
+        free(asn);
+        return obj->index;
+    }
+    obj = acalloc(1, sizeof(*obj));
+    if (NULL == obj) {
+        free(asn);
+        return -1;
+    }
+    /* If I directly use the buffer from libgeoip with
+         obj->asn = asn;
```

Comment:
This is because obj and entries in it uses arena functions so only memory allocated with that should be used.